xmppserver/src/main/java/org/jivesoftware/openfire/muc/MUCRoom.java

```diff
@@ -1308,6 +1308,48 @@ public void removeOccupant(@Nonnull final MUCOccupant occupant) {
         MUCEventDispatcher.occupantLeft(occupant.getOccupantJID(), occupant.getUserAddress(), occupant.getNickname());
     }
 
+    /**
+     * Clears the chat history of a room. Sends out message retraction stanzas to all occupants, for all the message in the chat history.
+     */
+    public void clearChatHistory() {
+        Iterator<Message> history = roomHistory.getMessageHistory();
+        while (history.hasNext()) {
+            Message originalMsg = history.next();
+            Message retractionMsg = new Message();
+
+            // Sets a unique ID for the retraction message by prefixing the original message ID with "retract-"
+            retractionMsg.setID("retract-" + originalMsg.getID());
+            // Sets the sender of the retraction message to be the chat room itself
+            retractionMsg.setFrom(selfOccupantData.getOccupantJID());
+            // Sets the recipient of the retraction message to be the chat room itself to send to all occupants
+            retractionMsg.setTo(new JID(getName(), getMUCService().getServiceDomain(), null).toBareJID());
+            retractionMsg.setType(Message.Type.groupchat);
+            // An XML element is added to the message to indicate that it is a retraction, with an attribute specifying
+            // the ID of the message being retracted
+            retractionMsg.addChildElement("retract", "urn:xmpp:message-retract:1").addAttribute(
+                "id",
+                new JID(getName(), getMUCService().getServiceDomain(), null).toBareJID()
+            );
+            // A fallback element is added to provide a fallback message for clients that do not support message retraction
+            retractionMsg.addChildElement("fallback", "urn:xmpp:fallback:0").addAttribute(
+                "for",
+                "urn:xmpp:message-retract:1"
+            );
+            retractionMsg.setBody("A request was received to retract a previous message as part of clearing the chat room history, but this is not supported by your client.");
```

Comment:
Please make this text a translatable text (using the i18n file). You can use `org.jivesoftware.util.LocaleUtils#getLocalizedString(java.lang.String)` to obtain the text from that file.

Reply:
Thanks, this is done now.